File: Examples/SnapinsSDKExample/app/src/javaImplDebug/java/com/salesforce/snapinssdkexample/activities/MainActivity.java
package com.salesforce.snapinssdkexample.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.salesforce.android.chat.core.ChatConfiguration;
import com.salesforce.android.chat.core.ChatCore;
import com.salesforce.android.chat.core.model.AvailabilityState;
import com.salesforce.android.service.common.analytics.ServiceAnalytics;
import com.salesforce.android.service.common.analytics.ServiceAnalyticsListener;
import com.salesforce.android.service.common.utilities.control.Async;
import com.salesforce.androidsdk.app.SalesforceSDKManager;
import com.salesforce.androidsdk.rest.ClientManager;
import com.salesforce.androidsdk.rest.RestClient;
import com.salesforce.snapinssdkexample.ChatLauncher;
import com.salesforce.snapinssdkexample.R;
import com.salesforce.snapinssdkexample.activities.settings.ChatSettingsActivity;
import com.salesforce.snapinssdkexample.utils.ServiceSDKUtils;

import java.util.Map;

/**
 * Main test activity supporting basic functionality:
 * <ul>
 * <li>Authentication (Logging in or out)</li>
 * </ul>
 */
public class MainActivity extends AppCompatActivity {
    private Button chatButton;
    private Button loginButton;
    private Button logoutButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);

        chatButton = findViewById(R.id.chat_launch_button);
        loginButton = findViewById(R.id.login_button);
        logoutButton = findViewById(R.id.logout_button);

        setSupportActionBar(toolbar);

        initializeServiceSDK();

        setupButtons();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_chat_settings:
                return startActivityFor(ChatSettingsActivity.class);
            case R.id.action_version_page:
                return startActivityFor(VersionActivity.class);
            case R.id.action_check_chat_agent_availability:
                return showLiveAgentChatAvailability();
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public boolean startActivityFor(Class<? extends AppCompatActivity> activityClass) {
        startActivity(new Intent(getApplicationContext(), activityClass));
        return true;
    }

    private Boolean showLiveAgentChatAvailability() {

        ChatConfiguration chatConfig = ServiceSDKUtils.getChatConfigurationBuilder(getApplicationContext()).build();

        // Create an agent availability client
        ChatCore.configureAgentAvailability(chatConfig).check()
                .onResult(new Async.ResultHandler<AvailabilityState>() {
                    @Override
                    public void handleResult(Async<?> async, @NonNull AvailabilityState availabilityState) {
                        // Display a toast when any agent availability state is changed
                        Toast.makeText(MainActivity.this,
                                String.format(getString(R.string.chat_availability_change_message), availabilityState.getStatus().toString()),
                                Toast.LENGTH_SHORT).show();
                    }
                });
        return true;
    }

    /**
     * Initializes configurations and listeners which should happen at startup.
     */
    private void initializeServiceSDK() {
        setupServiceSDKListeners();
    }

    /**
     * An example of how an application can listen to the Service SDK for analytic type events.
     */
    private void setupServiceSDKListeners() {
        // You can listen to user-driven events from the Snap-ins SDK using the ServiceAnalytics system.
        // Implement ServiceAnalyticsListener and add your listener to ServiceAnalytics to start receiving events.
        ServiceAnalytics.addListener(new ServiceAnalyticsListener() {
            @Override
            public void onServiceAnalyticsEvent(String behaviourId, Map<String, Object> eventData) {
                // Left blank intentionally
            }
        });
    }

    /**
     * Adds click listeners to the main activity buttons.
     */
    private void setupButtons() {
        chatButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                launchChat();
            }
        });
        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                login();
            }
        });
        logoutButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                logout();
            }
        });
    }

    /**
     * Launches Chat.
     */
    private void launchChat() {
        ChatLauncher chat = new ChatLauncher();
        chat.launchChat(this);
    }

    /**
     * Initiates user login process.
     */
    private void login() {
        SalesforceSDKManager.getInstance()
                .getClientManager()
                .getRestClient(this, new ClientManager.RestClientCallback() {
                    @Override
                    public void authenticatedRestClient(RestClient client) {
                        // left blank intentionally
                    }
                });
    }

    /**
     * Logs out authenticated users.
     */
    private void logout() {
        SalesforceSDKManager.getInstance().logout(this, true);
        Toast.makeText(this, "Logged out", Toast.LENGTH_SHORT).show();
    }

}

File: SnapinsSDKExample/app/src/kotlinImplDebug/java/com/salesforce/snapinssdkexample/activities/MainActivity.kt
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

package com.salesforce.snapinssdkexample.activities

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.salesforce.android.chat.core.ChatCore
import com.salesforce.android.chat.core.model.AvailabilityState
import com.salesforce.android.service.common.analytics.ServiceAnalytics
import com.salesforce.android.service.common.utilities.control.Async
import com.salesforce.androidsdk.app.SalesforceSDKManager
import com.salesforce.androidsdk.rest.RestClient
import com.salesforce.snapinssdkexample.ChatLauncher
import com.salesforce.snapinssdkexample.R
import com.salesforce.snapinssdkexample.activities.settings.ChatSettingsActivity
import com.salesforce.snapinssdkexample.utils.ServiceSDKUtils
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import kotlin.reflect.KClass

/**
 * Main test activity supporting basic functionality:
 * <ul>
 *     <li>Authentication (Logging in or out)</li>
 * </ul>
 */
class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        initializeServiceSDK()

        setupButtons()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_chat_settings -> startActivityFor(ChatSettingsActivity::class)
            R.id.action_version_page -> startActivityFor(VersionActivity::class)
            R.id.action_check_chat_agent_availability -> showLiveAgentChatAvailability()
            else -> super.onOptionsItemSelected(item)
        }
    }

    fun startActivityFor(activityClass: KClass<out AppCompatActivity>): Boolean {
        startActivity(Intent(applicationContext, activityClass.java))
        return true
    }

    private fun showLiveAgentChatAvailability(): Boolean {
        val chatConfig = ServiceSDKUtils.getChatConfigurationBuilder(applicationContext).build()

        // Create an agent availability client
        ChatCore.configureAgentAvailability(chatConfig).check()
                .onResult { _: Async<*>?, state: AvailabilityState ->
                    run {
                        // Display a toast when any agent availability state is changed
                        Toast.makeText(this,
                                String.format(getString(R.string.chat_availability_change_message), state.status.toString()),
                                Toast.LENGTH_SHORT).show()
                    }
                }

        return true
    }

    /**
     * Initializes configurations and listeners which should happen at startup.
     */
    private fun initializeServiceSDK() {
        setupServiceSDKListeners()
    }

    /**
     * An example of how an application can listen to the Service SDK for analytic type events.
     */
    private fun setupServiceSDKListeners() {
        // You can listen to user-driven events from the Snap-ins SDK using the ServiceAnalytics system.
        // Implement ServiceAnalyticsListener and add your listener to ServiceAnalytics to start receiving events.
        ServiceAnalytics.addListener({ behaviorId, eventData ->
            // Left blank intentionally
        })
    }

    /**
     * Adds click listeners to the main activity buttons.
     */
    private fun setupButtons() {
        chat_launch_button.setOnClickListener({ startChat() })
        login_button.setOnClickListener({ login() })
        logout_button.setOnClickListener({ logout() })
    }

    /**
     * Initializes Chat.
     */
    private fun startChat() {
        val chat = ChatLauncher()
        chat.launchChat(this)
    }

    /**
     * Initiates user login process.
     */
    private fun login() {
        SalesforceSDKManager.getInstance().clientManager.getRestClient(this) { client: RestClient ->
            run {
                // left blank intentionally
            }
        }
    }

    /**
     * Logs out authenticated users.
     */
    private fun logout() {
        SalesforceSDKManager.getInstance().logout(this, true)
        Toast.makeText(this, "Logged out", Toast.LENGTH_SHORT).show()
    }
}
